Run rule evaluation with timestamps precisely evaluation_interval apart

free opened the pull request:
Tests still to come.

I don't particularly like the fact that I'm relying on the ticker to trigger (1) soon after `interval` has passed and (2) never trigger early. I believe it's perfectly safe, I just don't like it.

free commented:
Hmm... I can't seem to find any Prometheus code/tests that allow for the use of fake clocks. Are there any you could point me to?

And if there aren't, should I:
(a) go ahead and vendor a package like https://github.com/jonboulle/clockwork; or 
(b) roll my own implementation that would also provide something along the lines of `NewTicker(interval, offset)`; or
(c) run some manual testing and declare it safe?

free commented:
BTW, here's the output of a manual test I ran with `--config.file=documentation/examples/prometheus.yml`, plus `first_rules.yml` defining the following rules:

```yml
groups:
  - name: example
    rules:
      - record: uptime
        expr: time() - process_start_time_seconds

      - record: iterations
        expr: prometheus_rule_group_iterations_total
```
![screenshot from 2018-05-29 14-34-49](https://user-images.githubusercontent.com/32070098/40659330-3e2b78e6-634e-11e8-9dfa-a93f651cd890.png)

You'll notice from the series' timestamps that I had Prometheus suspended twice: once from 1527596169+ to 1527596197; and once from 1527596223+ to 1527596285. Also notice how the recorded rule samples all have the exact same phase (timestamp modulo eval interval) as do the values of the `uptime` recorded rule, exactly as expected.

brian-brazil commented:
This seems more complicated than it needs to be. How about in `iter` taking the current time, rounding it down and then adding the offset?

free commented:
Assuming an interval of `1m`, a current time of 16:41:10 and an offset of 50 seconds, you'd end up with an eval timestamp of 16:41:50, which is in the future. You could first subtract the offset; then round; and finally add it back; but I'm not sure that's actually simpler.

That does give me some ideas on how to clean it up a bit though.

free commented:
With apologies for the number of commits, I've cleaned it up a little. I'm still using the same logic for advancing the timestamp, but I've dropped the type casts (`missing` is now a `time.Duration`, but no one should really care); and I've rewritten the eval time calculation so they don't use an `if` block anymore.

I could also use the eval time calculation directly within the `for` loop instead of the current increment based calculation, but it's going to be less efficient. Your call.

brian-brazil commented:
Thanks!

I wouldn't worry about the number of commits, it usually makes it easier to review for me and for a change like this I'm going to squash them anyway.
